libsupport/include/katana/ArrowVisitor.h

```diff
@@ -35,129 +22,166 @@ using is_list_type_patched = std::integral_constant<
+  return VisitArrowTraits<std::decay_t<CallerType>>::template Cast<ArrowType>(
+      std::forward<CallerType>(input));
+}
+
+/// ArrowDispatcher uses runtime arrow type information to downcast general
+/// arguments like arrow::Scalar to their specific runtime type like
+/// arrow::Int32Scalar and call a function on a visitor with the downcasted
+/// arguments along with their corresponding arrow types.
 ///
-/// In an ideal world, this is the *only* switch over Arrow types
-/// in the whole repo. Everything should either handle arrow data
-/// in a type-agnostic way, or use this interface to reach a
-/// type-aware template environment. It probably recreates
-/// functionality that exists elsewhere, notably in arrow::compute
+/// Effectively, given it converts
 ///
-/// A typical use case would involve creating a Visitor class
-/// and implementing methods to handle various types, eg:
-/// class Visitor {
-///   using ReturnType = void; // configurable
+///     ArrowDispatcher::Call(visitor, arrow::Scalar, arrow::Array)
 ///
-///   template <typename ArrowType, typename ArgumentType>
-///   katana::Result<ReturnType> Call(ArgumentType arg);
-/// };
-/// Arrow's type_traits.h offers tools to use SFINAE to differentiate
-/// types and write appropriate Call functions for each type
-
-// The VisitArrowInternalWrapper struct is used to keep track of the ArrowType for each argument
-// as we recursively cast each argument. The ArrowTypes are then used as
-// template arguments to the Call method of the Visitor class
-template <class... ArrowTypes>
-struct VisitArrowInternalWrapper {
-  // This function calls the visitor's templated Call method with the properly
-  // casted arguments and template parameters
-  template <class VisitorType, class... Processed, size_t... I>
-  static katana::Result<typename std::decay_t<VisitorType>::ReturnType>
-  VisitArrowInternalCall(
-      VisitorType&& visitor, std::tuple<Processed...>&& processed,
-      std::index_sequence<I...>) {
-    return visitor.template Call<ArrowTypes...>(std::get<I>(processed)...);
+/// into
+///
+///    visitor.Call<arrow::Int32Type, arrow::Int64Type>(arrow::Int32Scalar, arrow::Int64Array)
+///
+/// See VisitArrow for details on how this is typically used.
+struct ArrowDispatcher {
+  template <typename... ArrowTypes, typename Visitor, typename... Args>
+  static typename std::decay_t<Visitor>::ResultType Call(
+      Visitor&& visitor, Args&&... args) {
+    constexpr auto index = sizeof...(ArrowTypes);
+
+    if constexpr (index == sizeof...(Args)) {
+      // We have collected the runtime types of all our arguments, apply cast
+      // and call.
+      return visitor.template Call<ArrowTypes...>(
+          ApplyCast<ArrowTypes>(std::forward<Args>(args))...);
+    } else if constexpr (!std::decay_t<Visitor>::template CouldAccept<
+                             ArrowTypes...>()) {
+      return KATANA_ERROR(
+          katana::ErrorCode::ArrowError, "could not accept type id");
```

Comment:
I see how this avoids instantiating the type-specific visitor methods. What surprises me is that this is sufficient to keep the compiler from barfing. Wouldn't this function itself still be specialized for every type-specific case?
I guess that assumes that CouldAccept returns `true` until it 'sees' every argument, but it could short-circuit to prune the space earlier?
Is this intuition remotely on the right track?

Reply:
A curiosity more than a request, how hard would it be to derive the behavior of CouldAccept from types accepted by the methods of the visitor itself?

Reply:
This may be impossible (or not worth the time investment) but it would be awesome to have the visitor support specific types with `enable_if` and automatically prune if the `Call` function doesn't exist for the template specialization. Can def go in a future PR though.

Reply:
> I guess that assumes that CouldAccept returns true until it 'sees' every argument, but it could short-circuit to prune the space earlier?
> Is this intuition remotely on the right track?

Exactly. That's what I was trying to convey with the name `CouldAccept`. But yes, `CouldAccept` needs to be a  incremental and overapproximates `visitor.Call` and we get the main effect if it short-circuits.

> how hard would it be to derive the behavior of CouldAccept from types accepted by the methods of the visitor itself?

It is not possible to know if a function exists without first providing arguments due to function overloading, so I don't think it is possible to derive the information we need solely from the definition of `visitor.Call`.

This does remind me: I also considered providing a catch all case in `VisitArrow` itself, so that visitors can avoid making themselves total functions like we do now:

```cpp
struct Visitor {
  template<typename ....>
  std::enable_if_t<P, ResultType> Call(....) {}

  template<typename ....>
  std::enable_if_t<!P, ResultType> Call(....) {
    // handle error case
  }
};
```

but I thought it might be better for visitors to explicitly handle these cases to force developers to think about these cases and provide better error messages. But this gets somewhat ugly as we increase the number parameters.

Thoughts?